File: client/src/modules/nexotype/schemas/omics/gene.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const GeneSchema = z.object({
  id: z.number(),
  organism_id: z.number().int(),
  hgnc_symbol: z.string().min(1).max(50),
  ensembl_gene_id: z.string().min(1).max(50),
  chromosome: z.string().min(1).max(10),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateGeneSchema = z.object({
  organism_id: z.number().int(),
  hgnc_symbol: z.string().min(1).max(50),
  ensembl_gene_id: z.string().min(1).max(50),
  chromosome: z.string().min(1).max(10),
});

export const UpdateGeneSchema = z.object({
  organism_id: z.number().int().optional(),
  hgnc_symbol: z.string().min(1).max(50).optional(),
  ensembl_gene_id: z.string().min(1).max(50).optional(),
  chromosome: z.string().min(1).max(10).optional(),
});

// ==========================================
// Types
// ==========================================

export type Gene = z.infer<typeof GeneSchema>;
export type CreateGene = z.infer<typeof CreateGeneSchema>;
export type UpdateGene = z.infer<typeof UpdateGeneSchema>;

// ==========================================
// Response Types
// ==========================================

export type GeneResponse = {
  success: boolean;
  data?: Gene;
  error?: string;
};

export type GenesResponse = {
  success: boolean;
  data?: Gene[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/omics/gene_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class GeneCreate(BaseModel):
    """Schema for creating a new gene"""
    organism_id: int
    hgnc_symbol: str
    ensembl_gene_id: str
    chromosome: str


class GeneUpdate(BaseModel):
    """Schema for updating a gene"""
    organism_id: int = None
    hgnc_symbol: str = None
    ensembl_gene_id: str = None
    chromosome: str = None


# ==========================================
# Response Schemas
# ==========================================

class GeneDetail(BaseModel):
    """Schema for gene details"""
    id: int
    organism_id: int
    hgnc_symbol: str
    ensembl_gene_id: str
    chromosome: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class GeneListResponse(BaseModel):
    """Response schema for listing genes"""
    success: bool
    data: list[GeneDetail] | None = None
    count: int = 0
    error: str | None = None


class GeneResponse(BaseModel):
    """Response schema for single gene operations"""
    success: bool
    data: GeneDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
